os/mm/mm_heap/mm_check_heap_corruption.c

```diff
@@ -164,100 +168,117 @@ int mm_check_heap_corruption(struct mm_heap_s *heap)
 		 * Retake the semaphore for each region to reduce latencies
 		 */
 
-		if (!aborted
 #if defined(CONFIG_BUILD_FLAT) || defined(__KERNEL__)
-			&& !up_interrupt_context()
+		if (!abort_mode && !up_interrupt_context()) 
 #endif
-		) {
+		{
 			mm_takesemaphore(heap);
 		}
 
-		prev = NULL;
-		node = heap->mm_heapstart[region];
-		next = (struct mm_allocnode_s *)((char *)node + node->size);
+		mfdbg("Heap start = 0x%08x end = 0x%08x\n", heap->mm_heapstart[region], heap->mm_heapend[region]);
 
-		for (; node < heap->mm_heapend[region]; prev = node, node = next, next = (struct mm_allocnode_s *)((char *)next + next->size)) {
-			if (prev && prev->size != MM_PREV_NODE_SIZE(node)) {
-				mfdbg("#########################################################################################\n");
-				mfdbg("ERROR: Heap node corruption detected\n");
-				dump_node(prev, TYPE_OVERFLOWED);
-				dump_node(node, TYPE_CORRUPTED);
-				mfdbg("#########################################################################################\n");
-				if (!aborted
-#if defined(CONFIG_BUILD_FLAT) || defined(__KERNEL__)
-					&& !up_interrupt_context()
-#endif
-				) {
-					mm_givesemaphore(heap);
+		if (((struct mm_allocnode_s *)(heap->mm_heapstart[region]))->size != SIZEOF_MM_ALLOCNODE) {
+			mfdbg("#########################################################################################\n");
+			mfdbg("ERROR: Heap corruption detected in mm_heapstart node.\n");
+			mfdbg("ERROR: Check for overflow during backward traversal of heap\n");
+			mfdbg("=========================================================================================\n");
+			dump_node(heap->mm_heapstart[region], "HEAP START NODE");
+			iscorrupt_f = true;
+			start_corrupt = (uint32_t)(heap->mm_heapstart[region]);
+			mfdbg("=========================================================================================\n");
+		} else {
+			/* Forward traversal of heap */
+			for (prev = heap->mm_heapstart[region], node = (struct mm_allocnode_s *)((char *)prev + prev->size);
+					node <= heap->mm_heapend[region];
+					prev = node, node = (struct mm_allocnode_s *)((char *)node + node->size)) {
```

Comment:
what if node->size becomes 0 due to corruption? I think this loop runs forever then. Could you please add a check for that?

Reply:
Oh.. you are right!!
Fixed. Thank you.